tests/graphql_spec.rs

```diff
@@ -23,16 +23,65 @@ mod graphql_mock;
 
 static INIT: Once = Once::new();
 
+#[derive(Debug, Clone, PartialEq)]
+enum Tag {
+  ClientSDL,
+  ServerSDL,
+  MergedSDL,
+}
+
+#[derive(Debug, Clone)]
+struct Source {
+  sdl: String,
+  tag: Tag,
+}
+
 #[derive(Debug, Default, Setters)]
 struct GraphQLSpec {
   path: PathBuf,
-  client_sdl: String,
-  server_sdl: Vec<String>,
-  merged_server_sdl: String,
+  sources: Vec<Source>,
   sdl_errors: Vec<SDLError>,
   test_queries: Vec<GraphQLQuerySpec>,
 }
 
+impl GraphQLSpec {
+  fn client_sdl(mut self, sdl: String) -> Self {
+    self.sources.push(Source { sdl, tag: Tag::ClientSDL });
+    self
+  }
+
+  fn server_sdl(mut self, sdl: Vec<String>) -> Self {
+    for s in sdl {
+      self.sources.push(Source { sdl: s, tag: Tag::ServerSDL });
+    }
+    self
+  }
+
+  fn merged_server_sdl(mut self, sdl: String) -> Self {
+    self.sources.push(Source { sdl, tag: Tag::MergedSDL });
+
+    self
+  }
+
+  fn find_source(&self, tag: Tag) -> String {
+    self
+      .get_sources(tag)
+      .collect::<Vec<String>>()
+      .first()
+      .unwrap()
+      .to_string()
+  }
+
+  fn get_sources(&self, tag: Tag) -> impl Iterator<Item = String> {
+    self
+      .sources
+      .clone()
+      .into_iter()
+      .filter(move |s| s.tag == tag)
+      .map(|s| s.sdl)
+  }
```

Comment:
yeah, good start, and you can improve it even more!
- I think we can get do just `self.source.iter()` to not clone anything from the beginning - we'll have to work with references there but this is even better
- for doing `.first()` we don't even need to collect anything as it can be replaced by `.next()` on iterator itself
- returning String in some cases could be overcopying things and if you can use  &str it's better use it

Reply:
> for doing .first() we don't even need to collect anything as it can be replaced by .next() on iterator itself

done. nice catch!

> I think we can get do just self.source.iter() to not clone anything from the beginning - we'll have to work with references there but this is even better

it does not work because `Source` does not implement the `Copy` trait (I've tried to implement it myself but gave up after a lot of work).

> returning String in some cases could be overcopying things and if you can use &str it's better use it

while this makes a lot of sense, I'm struggling to implement it; in the end, I'd have way more `.clone()` to have it working

Reply:
this code could look like this to minimize cloning

```rust
  fn find_source(&self, tag: Tag) -> Option<&str> {
    self.get_sources(tag).next()
  }

  fn get_sources(&self, tag: Tag) -> impl Iterator<Item = &str> {
    self
      .sources
      .iter()
      .filter(move |s| s.tag == tag)
      .map(|s| s.sdl.as_str())
  }
  ```

Reply:
done, @meskill, it worked pretty well, thank you :smile:

Reply:
I did not succeed on changing `find_source` signature, ended up having issues with it but for `get_sources` it worked and we're not cloning anymore.

Reply:
yeah it requires changes in usage as well in that case, anyway it is minor thing no need to spend time on this